### PR TITLE
Graph selection state management and history bug fixes

### LINKED ISSUE #203: reactive repaint bug - brush selection
If you load the app, brush select some cells in the main graph, and then resize (change width) the window, the lasso will not repaint.

### LINKED ISSUE #622: main graph zoom & lasso don't work together
there are a number of significant bugs in the new lasso select function, manifest when the user also uses the zoom/pan functionality.   This all appears to have been introduced in the 0.6 release with Lasso.

Cases I found with a bit of quick testing:

1. Lasso doesn't track pan/zoom coordinate system changes.   To reproduce:
    1. load a dataset
    2. lasso select a region
    3. Click the pan/zoom control to switch to pan/zoom mode

         > Bug (A) - the lasso disappears when you switch to pan/zoom mode

    4. Pan the graph 
    5. Click the lasso button to switch back to lasso select mode

          > Bug (B) - the lasso reappears, but it is in the wrong location (it did not translate with the pan)

    You can see the same thing when you zoom (rather than pan).   The coordinate system changes of the underlying graph are not linked to the SVG (lasso) layer.

2. Selection state is lost when switching between Lasso and Pan/Zoom modes.  To reproduce:
    1. load dataset
    2. lasso select a region
    3. switch to Pan/Zoom mode (this manifests bug (A))

          > Bug (C) - the selection state seems to be reset

    4. Click the Lasso button to switch to Lasso mode

          > Bug (D) - the lasso repaints, but it doesn't match the selection state (ie, cells outside the lasso are still selected)

    So in this case, the selection state gets disconnected from the lasso paint...

### LINKED ISSUE #663: Reset/Undo do not correctly update lasso state
The lasso paint state is not tied to the redux state, so it does not update correctly when the underlying redux store updates.    For example, as a result of a Reset or Undo/Redo.

### LINKED ISSUE #692: multiple clicks on the same color-by icon will push multiple undo points in the app history
Click repeatedly on any (single) color-by button, and multiple history undo points are created.  These should be consolidated.

### PR DESCRIPTION
**Pull request by bkmartinjr**
Changes in this PR:
* Graph selection (rect and poly) now synchronized with graph pan/zoom state.  Fixes #622
* Both poly and rect selection now work correctly with undo/redo.   As part of this, graph selection state and actions refactored to store state in the redux store, and to be responsive to that state.  Fixes #663 
* Graph selection outline repaints with view change.  Fixes #203 
* Undoable reducer enhanced to support more complex multi-action compositions, which are needed to fully capture async or multi-action behavior of the graph selection process. This includes:
    * better separation of concerns between the undoable reducer core and per-action filters that driver the behavior of undoable.
    * a new simple state machine (`util/statemachine`) and is configuration/use in the undoable action filter.
* Graph brush (rect) selection works again, fixing several regressions since lasso introduced.
* Color-by buttons no longer create multiple history checkpoints when clicked repeatedly.  Fixes #692

UI Changes:
* Toggling between pan/zoom and selection on the main embedding graph no longer resets the selection.
* Several unexpected/undesired undo checkpoints removed, improving undo/redo consistency
* window resizing correctly repaints graph selection UI

### CODE REVIEW

File: client/src/components/graph/setupLasso.js [hunk 1]
```diff
@@ -114,6 +115,22 @@ const Lasso = () => {
...
       }
     };
+
+    lasso.move = polygon => {
+      if (polygon !== lassoPolygon || polygon.length !== lassoPolygon.length) {
+        lasso.reset();
+
+        lassoPolygon = polygon;
+        lassoPath = g
+          .append("path")
```
Comment (csweaver) [inline client/src/components/graph/setupLasso.js]: Please add to facilitate smoke testing:
```suggestion
          .append("path")
          .attr("data-testid", "lasso-element")
```

Comment (bkmartinjr) [reply client/src/components/graph/setupLasso.js]: added.

File: client/src/components/geneExpression/index.js [hunk 2]
```diff
@@ -136,22 +137,22 @@ class GeneExpression extends React.Component {
...
+            return keepAroundErrorToast("That gene already exists");
+          }
+          if (world.varAnnotations.col("name").indexOf(gene) === undefined) {
+            return keepAroundErrorToast(
+              `${gene} doesn't appear to be a valid gene name.`
+            );
+          }
+          return dispatch(actions.requestUserDefinedGene(gene));
+        })
+      ).then(() => dispatch({ type: "bulk user defined gene complete" }));
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply client/src/components/geneExpression/index.js]: Can't hurt, but I don't think it is strictly necessary as the upstream errors are all handled, and inside requestUserDefinedGene().   But as future-proofing, it seems prudent.

[the remark this answers is not in the document]
Comment (bkmartinjr) [reply client/src/components/geneExpression/index.js]: added

**Pull request merged**